samcli/lib/generated_sample_events/events.py

```diff
@@ -33,63 +33,106 @@ def __init__(self):
+            val = values_to_sub.get(tag)
+            values_to_sub[tag] = self.transform_val(properties, val)
+            if properties.get("children") is not None:
+                children = properties.get("children")
+                for child_tag, child_properties in children.items():
+                    child_val = self.transform_val(child_properties, val)
+                    values_to_sub[child_tag] = child_val
         return values_to_sub
 
-    def url_encode(self, value):
+    def transform_val(self, properties, val):
         """
-        url encodes the value passed in
+        transform (if needed) given val with given properties
 
         Parameters
         ----------
-        value: string
-            the value that needs to be encoded in the json
+        properties: dict
+            set of properties to be used for transformation
+        val: string
+            the value to undergo transformation
         Returns
         -------
-        string: the url encoded value
+        transformed
+            the transformed value
+        """
+        transformed = val
+
+        # encode if needed
+        encoding = properties.get("encoding")
+        if encoding is not None:
+            transformed = self.encode(encoding, transformed)
+
+        # hash if needed
+        hashing = properties.get("hashing")
+        if hashing is not None:
+            transformed = self.hash(hashing, transformed)
+
+        return transformed
+
+    def encode(self, encoding_scheme, val):
         """
+        encodes a given val with given encoding scheme
+
+        Parameters
+        ----------
+        encoding_scheme: string
+            the encoding scheme
+        val: string
+            the value to be encoded
+        Returns
+        -------
+        encoded: string
+            the encoded value
+        """
+        if encoding_scheme == "url":
+            return url_quote(val)
+
+        # base64 utf8
+        if encoding_scheme == "base64":
+            return base64.b64encode(val.encode("utf8")).decode("utf-8")
 
-        return quote(value)
+        # returns original val if encoding_scheme not recognized
+        return val
 
-    def base64_utf_encode(self, value):
+    def hash(self, hashing_scheme, val):
         """
-        base64 utf8 encodes the value passed in
+        hashes a given val using given hashing_scheme
 
         Parameters
         ----------
-        value: string
-            value that needs to be encoded in the json
+        hashing_scheme: string
+            the hashing scheme
+        val: string
+            the value to be hashed
         Returns
         -------
-        string: the base64_utf encoded value
+        hashed: string
+            the hashed value
         """
+        if hashing_scheme == "md5":
+            return str_checksum(val)
 
-        return base64.b64encode(value.encode("utf8")).decode("utf-8")
+        # returns original val if hashing_scheme not recognized
```

Comment:
It's consistent with previous behavior, but is this what we want to do? It seems to me that at minimum we should log a warning, and perhaps we should raise an exception in this case. For our pre-baked events, this exception should never occur and we would want an explicit failure and for someone to raise an issue if it does.

Reply:
yep it makes sense to raise an exception. Will update this part.